added api page for CSSNestedDeclaration and the style peroperty and a… (#36430)
mdn/content: files/en-us/web/api/cssnesteddeclarations/index.md

---
title: CSSNestedDeclarations
slug: Web/API/CSSNestedDeclarations
page-type: web-api-interface
browser-compat: api.CSSNestedDeclarations
---

{{APIRef("CSSOM")}}

> The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) allows the nesting of `CSSRules` within each other.
> 
> With the nested declaration rule the rules can nested and are all grouped together to retain the order.

Reviewer: A bit confusing if you don't already know what it means. How about.

```suggestion
The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) is used to group nested {{domxref("CSSRules"}}.

The interface allows the [CSS Object Model (CSSOM](/en-US/docs/Web/API/CSS_Object_Model) to mirror the structure of CSS documents with nested CSS rules, and ensure that rules are parsed and evaluated in the order that they are declared.
```

---
title: CSSNestedDeclarations
slug: Web/API/CSSNestedDeclarations
page-type: web-api-interface
browser-compat: api.CSSNestedDeclarations
---

{{APIRef("CSSOM")}}

> The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) is used to group nested {{domxref("CSSRules")}}.

Reviewer: ```suggestion
The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) is used to group nested {{domxref("CSSRule")}}s.
```

## Instance methods

_No specific methods; inherits methods from its ancestor {{domxref("CSSRule")}}._

## Examples

The CSS below includes a selector `.foo` with a three of rules that define it:

- The first is a {{domxref("CSSStyleRule")}} object which can be returned via `document.styleSheets[0].cssRules[0]` and is equivalent to `background-color: silver`
- The second is a {{domxref("CSSMediaRule")}} object which can be returned via `document.styleSheets[0].cssRules[0].cssRules[0]` and is equivalent to `@media (screen)`
  - The `CSSMediaRule` object contains a `CSSNestedDeclaration` object which can be returned via `document.styleSheets[0].cssRules[0].cssRules[0].cssRules[0]` and is equivalent to `color: tomato`
- The third is a `CSSNestedDeclaration` object which can be returned via `document.styleSheets[0].cssRules[0].cssRules[1]` and is equivalent to `color: black`

```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

The following JavaScript code returns a CSSNestedDeclarations object.

```js
let myRules = document.styleSheets[0].cssRules;
console.log(myRules[0].cssRules[1]); // [object CSSNestedDeclarations]
```

> 

Reviewer: 1. I don't know what "three of rules" means, assume typo
2. I think the bit after "The following JavaScript code returns a CSSNestedDeclarations object." is now unnecessary/accidental leftover
3. I think this is on the right track, but these rules are CSSOM rules and even though the mapping between stylesheet and the CSSOM should be pretty much 1:1, better to present them as stylesheet then representation
 
I can't make this a suggestion because it breaks the suggestions system, but here's the proposed text.

<pre>
The CSS below includes a selector `.foo` that contains three nested rules.

```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

This is represented by a number of JavaScript objects in the [CSS Object Model](/en-US/docs/Web/API/CSS_Object_Model):

- A {{domxref("CSSStyleRule")}} object that represents the `background-color: silver` rule.
  This can be returned via `document.styleSheets[0].cssRules[0]`.
- A {{domxref("CSSMediaRule")}} object  that represents the `@media (screen)` rule, and which can be returned via `document.styleSheets[0].cssRules[0].cssRules[0]`.
   - The `CSSMediaRule` object contains a `CSSNestedDeclaration` object which represents the  `color: tomato` rule nested by the `@media (screen)` rule.
     This can be returned via `document.styleSheets[0].cssRules[0].cssRules[0].cssRules[0]`.
- The final rule is a `CSSNestedDeclaration` object that represents the `color: black` rule in the stylesheet, and which can be returned via `document.styleSheets[0].cssRules[0].cssRules[1]`.

  Note that all top-level styles after the first `CSSNestedDeclaration` must also be represented as `CSSNestedDeclaration` objects in order to follow the [CSS nested declarations rule](/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#nested_declarations_rule)
</pre>

Should we mention the actual representation of the rules as CSSStyleRules - i.e. it is a  `CSSNestedDeclaration` that contains a CSSNestedDeclarations (I think)? This is kind of why I was thinking it might be better to literally construct the CSSOM rule in code rather than show how the rule is parsed to

Reviewer: PS can you invite me to your repo. I would have merged this myself had you done so.

Reviewer 2: > I can't make this a suggestion because it breaks the suggestions system

@hamishwillee Add one more backtick to the open/close suggestion fence when there’s another code fence inside, and it’ll work just fine.

`````
````suggestion
```css
body { color: red }
```
````
`````

I wrapped this one in five backticks to show :)

Reviewer: THanks very very much @pepelsbey - that's going to come in handy! Updated above.

---
title: CSSNestedDeclarations
slug: Web/API/CSSNestedDeclarations
page-type: web-api-interface
browser-compat: api.CSSNestedDeclarations
---

{{APIRef("CSSOM")}}

> The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) allows the nesting of `CSSRules` within each other. This was added to the [CSS Object Model (CSSOM)](/en-US/docs/Web/API/CSS_Object_Model) to fix an issue where nested rules could be parsed by the browser in the wrong order.

Reviewer: Having a sentence like "This was added to the [CSS Object Model (CSSOM)](/en-US/docs/Web/API/CSS_Object_Model) to fix an issue where nested rules could be parsed by the browser in the wrong order." seems very odd. 
This is a compat issue that will one day be irrelevant, so perhaps.

```suggestion
The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) allows the nesting of `CSSRules` within each other.

> [!NOTE]
> [Browser versions](#browser_compatibility) that implementations that do not support this interface may parse nested rules in the wrong order.
```

Author: Thank you, this makes total sense

{{InheritanceDiagram}}

## Instance properties

_Inherits properties from its ancestor {{domxref("CSSRule")}}._

- {{domxref("CSSNestedDeclarations.style")}} {{ReadOnlyInline}}
  - : Returns the values of the nested rules.

## Instance methods

_No specific methods; inherits methods from its ancestor {{domxref("CSSRule")}}._

## Examples

> This stylesheet contains a rule followed by a nested [@media](/en-US/docs/Web/CSS/@media) CSS at-rule followed by another rule.

Reviewer: ```suggestion
This stylesheet contains a rule, followed by a nested [@media](/en-US/docs/Web/CSS/@media) CSS at-rule, followed by another rule.
```

```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

### Without CSSNestedDeclarations

Without `CSSNestedDeclarations` the CSS is is parsed in the wrong order, meaning the text color will be `tomato` and not `black` as the author intended.

```css
.foo {
  background-color: silver;
  color: black;
  @media (screen) {
    color: tomato;
  }

> }

Reviewer: I don't love this example. As above this shows a bug and how it is solved, not how I use this interface.

I would show it as code that constructs the form

```
```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

Or if that is too hard, you could show code that prints out this whole structure and show the log like thi:

```
↳ CSSStyleRule
  .style
    - background-color: silver
  ↳ CSSMediaRule
    ↳ CSSNestedDeclarations
      .style (CSSStyleDeclaration, 1) =
      - color: tomato
  ↳ CSSNestedDeclarations
    .style (CSSStyleDeclaration, 1) =
      - color: black
   ```
   
   If mentioned at all, the "without nested declarations", is a tiny note.

Author: I have removed this altogether as the other APIs don't have these at all.

```

### With CSSNestedDeclarations

With `CSSNestedDeclarations` the CSS is is parsed correctly.

```js

> let myRules = document.styleSheets[0].cssRules;

Reviewer: As above, this really doesn't tell me anything because I don't see what the types are.

Author: See https://github.com/mdn/content/pull/36430#discussion_r1818224250

Reviewer: See my response https://github.com/mdn/content/pull/36430#discussion_r1819838617

What do you think a reader can learn from this? I don't particularly might what you show, but it has to show something related to the API we are demonstrating, and really this does not.

If I had to do this I would either have a live example that shows the structure, or add pseudo code. Assuming the second, then something like. 

---

Consider the following

```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

This would be parsed into a structure similar to the one shown below. Note that blah blah. 

**[Note, that I am not sure this is correct - I just copied it from your original. Probably it isn't - ie. I don't think the last declaration is nested]**
```
↳ CSSStyleRule
  .style
    - background-color: silver
  ↳ CSSMediaRule
    ↳ CSSNestedDeclarations
      .style (CSSStyleDeclaration, 1) =
      - color: tomato
  ↳ CSSNestedDeclarations
    .style (CSSStyleDeclaration, 1) =
      - color: black
```

Author: Yes it is, as this is how the order of the rules are retained.
- https://codepen.io/CodeRedDigital/pen/zYgdaWy
- https://web.dev/blog/css-nesting-cssnesteddeclarations

Reviewer: > What do you think a reader can learn from this? I don't particularly might what you show, but it has to show something related to the API we are demonstrating, and really this does not.

Still not answered.

Reviewer: What I would do here is write code that creates this structure. I'd do it for you but your repo is locked for edits. So something like:

---

This example shows how to construct the following stylesheet using the CSSDOM API.

```css
.foo {
  background-color: silver;
  @media (screen) {
    color: tomato;
  }
  color: black;
}
```

[Insert the code here]

Note that in order to preserve the parsing order, all the rules before nesting are handled as top-level `CSSRules`, while any top level rules after nesting are represented as `CSSNestedDeclarations`.
That's why the `color-black` is inside a nested declaration even though it is a top level declaration in the original document.

----

Note that point at the end that you gave me in the other note - this is relevant to usage of this API.

Author: I am not sure that this is correct, either.
The `CSSNestedDeclarations` API only returns that the there is an object of type `CSSNestedDeclarations` which in turn are `cssRules` with `style`s
https://codepen.io/CodeRedDigital/pen/vYozyrG?editors=0112

Although I believe I has managed to explain this better now

console.log(myRules[0].cssRules[1]); // [object CSSNestedDeclarations]
```

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See Also

- {{domxref("CSSNestedDeclarations.style")}}

> - [The Nested Declarations Rule](https://drafts.csswg.org/css-nesting-1/#nested-declarations-rule)

Reviewer: Why not  link to /en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#nested_declarations_rule ?

Author: Done

---
title: CSSNestedDeclarations
slug: Web/API/CSSNestedDeclarations
page-type: web-api-interface
browser-compat: api.CSSNestedDeclarations
---

{{APIRef("CSSOM")}}

The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) allows the nesting of `CSSRules` within each other.

> > [!NOTE] > [Browser versions](#browser_compatibility) that implementations that do not support this interface may parse nested rules in the wrong order.

Reviewer: ```suggestion
> [!NOTE]
> [Browser versions](#browser_compatibility) that implementations that do not support this interface may parse nested rules in the wrong order.
```

{{APIRef("CSSOM")}}

The **`CSSNestedDeclarations`** interface of the [CSS Rule API](/en-US/docs/Web/API/CSSRule) allows the nesting of `CSSRules` within each other.

> > [!NOTE] > [Browser versions](#browser_compatibility) that implementations that do not support this interface may parse nested rules in the wrong order.

Reviewer: ```suggestion
> [!NOTE]
> [Browser versions](#browser_compatibility) that implementations that do not support this interface may parse nested rules in the wrong order.
```
